cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java

```diff
@@ -423,8 +427,13 @@ public ArtifactDetail addArtifact(Id.Artifact artifactId, File artifactFile,
                                     @Nullable Set<PluginClass> additionalPlugins,
                                     Map<String, String> properties) throws Exception {
     // check that the current user has write privileges on the namespace
-    authorizerInstantiatorService.get().enforce(artifactId.getNamespace().toEntityId(),
-                                                SecurityRequestContext.toPrincipal(), Action.WRITE);
+    // Skip authorization enforcement for system user. This is so the SystemArtifactLoader thread can add system
+    // artifacts.
+    Principal principal = SecurityRequestContext.toPrincipal();
+    NamespaceId namespace = artifactId.getNamespace().toEntityId();
+    if (!(Principal.SYSTEM.equals(principal) && NamespaceId.SYSTEM.equals(namespace))) {
```

Comment:
It's just confusing that this is here, doesn't seem like the method should be responsible for whether it needs authorize or not. How about we pull the add artifact logic into a private method without auth and have this method call that new private method and also have the addSystemArtifacts method call that as well.